service/history/mutableStateBuilder.go

```diff
@@ -246,6 +240,67 @@ func (e *mutableStateBuilder) CloseUpdateSession() (*mutableStateSessionUpdates,
 	return updates, nil
 }
 
+func (e *mutableStateBuilder) assignEventIdToBufferedEvents() {
+	newCommittedEvents := e.hBuilder.history
+
+	for i, event := range newCommittedEvents {
+		if event.GetEventId() != bufferedEventID {
+			continue
+		}
+
+		eventID := e.executionInfo.NextEventID
+		event.EventId = common.Int64Ptr(eventID)
+		e.executionInfo.NextEventID++
+
+		switch event.GetEventType() {
+		case workflow.EventTypeActivityTaskStarted:
+			// need to update startedEventId in mutableState if it is still there
+			scheduledEventId := event.ActivityTaskStartedEventAttributes.GetScheduledEventId()
+			if ai, ok := e.GetActivityInfo(scheduledEventId); ok {
+				ai.StartedID = eventID
+				e.updateActivityInfos = append(e.updateActivityInfos, ai)
+			}
+			// update subsequent buffered events in this batch that have reference to the startedEventId
+			for j := i + 1; j < len(newCommittedEvents); j++ {
```

Comment:
I think we should be able to update all events in a single pass.  I'm not sure this N^2 algorithm is needed here.